File: openspec/changes/archive/ux-plugins-control-panel/example_client.js
/**
 * Example JavaScript client for Roamin Control API
 *
 * Usage:
 *   node example_client.js
 *   // or in browser:
 *   // <script src="example_client.js"></script>
 *   // const client = new RoaminControlClient();
 *   // await client.getStatus().then(s => console.log(s));
 */

class RoaminControlClient {
  constructor(baseUrl = "http://127.0.0.1:8765") {
    this.baseUrl = baseUrl;
    this.apiKey = process.env?.ROAMIN_CONTROL_API_KEY || "";
  }

  /**
   * Helper to make authenticated requests
   */
  async request(method, path, body = null) {
    const options = {
      method,
      headers: {
        "Content-Type": "application/json",
      },
    };

    if (this.apiKey) {
      options.headers["x-roamin-api-key"] = this.apiKey;
    }

    if (body) {
      options.body = JSON.stringify(body);
    }

    const response = await fetch(`${this.baseUrl}${path}`, options);
    if (!response.ok) {
      throw new Error(`${method} ${path} returned ${response.status}`);
    }
    return await response.json();
  }

  /**
   * Get system status
   */
  async getStatus() {
    return this.request("GET", "/status");
  }

  /**
   * List available models
   */
  async listModels() {
    return this.request("GET", "/models");
  }

  /**
   * List installed plugins
   */
  async listPlugins() {
    return this.request("GET", "/plugins");
  }

  /**
   * Install a plugin from file or URL
   * @param {string} source - "file" or "url"
   * @param {string} value - path or HTTPS URL
   */
  async installPlugin(source, value) {
    return this.request("POST", "/plugins/install", { source, value });
  }

  /**
   * Uninstall a plugin
   * @param {string} pluginId - Plugin identifier
   */
  async uninstallPlugin(pluginId) {
    return this.request("DELETE", `/plugins/${pluginId}`);
  }

  /**
   * Perform plugin action (enable, disable, restart)
   * @param {string} pluginId - Plugin identifier
   * @param {string} action - "enable", "disable", or "restart"
   */
  async pluginAction(pluginId, action) {
    return this.request("POST", `/plugins/${pluginId}/action`, { action });
  }

  /**
   * Get task history
   * @param {number} limit - Max tasks to return
   * @param {string} status - Filter by status (optional)
   */
  async getTaskHistory(limit = 100, status = null) {
    let path = `/task-history?limit=${limit}`;
    if (status) {
      path += `&status=${status}`;
    }
    return this.request("GET", path);
  }

  /**
   * Connect to WebSocket event stream
   * @param {function} onEvent - Callback for each event
   * @param {function} onError - Callback for errors
   */
  listenEvents(onEvent, onError) {
    const wsUrl = "ws://127.0.0.1:8765/ws/events";
    console.log(`[*] Connecting to ${wsUrl}`);

    const ws = new WebSocket(wsUrl);

    ws.onopen = () => {
      console.log("[+] Connected! Listening for events...");
    };

    ws.onmessage = (event) => {
      try {
        const data = JSON.parse(event.data);
        console.log(`[EVENT] ${data.type}:`, data.data);
        if (onEvent) onEvent(data);
      } catch (err) {
        console.error("[!] Failed to parse event:", err);
      }
    };

    ws.onerror = (event) => {
      console.error("[!] WebSocket error:", event);
      if (onError) onError(event);
    };

    ws.onclose = () => {
      console.log("[-] Disconnected");
    };

    return ws;
  }
}

/**
 * Demo: Connect to API and run sample commands
 */
async function main() {
  const client = new RoaminControlClient();

  try {
    // Get status
    console.log("[*] Fetching status...");
    const status = await client.getStatus();
    console.log("[+] Status:", JSON.stringify(status, null, 2));

    // List models
    console.log("\n[*] Fetching models...");
    const models = await client.listModels();
    console.log("[+] Models:", JSON.stringify(models, null, 2));

    // List plugins
    console.log("\n[*] Fetching plugins...");
    const plugins = await client.listPlugins();
    console.log("[+] Plugins:", JSON.stringify(plugins, null, 2));

    // Get task history
    console.log("\n[*] Fetching task history...");
    const tasks = await client.getTaskHistory(10);
    console.log("[+] Tasks:", JSON.stringify(tasks, null, 2));

    // Listen for events (run for 5 seconds as demo)
    console.log("\n[*] Listening for events (5s timeout)...");
    const ws = client.listenEvents(
      (event) => {
        // Handle individual events here
      },
      (error) => {
        console.error("[!] Event listener error:", error);
      }
    );

    // Close after 5 seconds
    setTimeout(() => {
      ws.close();
    }, 5000);
  } catch (error) {
    console.error("[!] Error:", error.message);
    console.error("    Is Control API running at http://127.0.0.1:8765?");
  }
}

// Run demo if this is the main module (Node.js)
if (typeof module !== "undefined" && require.main === module) {
  main();
}

// Export for use as module
if (typeof module !== "undefined") {
  module.exports = RoaminControlClient;
}

File: openspec/changes/archive/ux-plugins-control-panel/example_client.py
#!/usr/bin/env python3
"""
Example Python client for Roamin Control API.

Usage:
    python example_client.py
"""

from __future__ import annotations

import asyncio
import json
import os

import aiohttp
import websockets


class RoaminControlClient:
    """Simple async client for Roamin Control API."""

    def __init__(self, base_url: str = "http://127.0.0.1:8765"):
        self.base_url = base_url
        self.api_key = os.environ.get("ROAMIN_CONTROL_API_KEY")
        self.headers = {}
        if self.api_key:
            self.headers["x-roamin-api-key"] = self.api_key

    async def get_status(self) -> dict:
        """Fetch system status."""
        async with aiohttp.ClientSession() as session:
            async with session.get(
                f"{self.base_url}/status",
                headers=self.headers,
            ) as resp:
                return await resp.json()

    async def list_models(self) -> dict:
        """List available models."""
        async with aiohttp.ClientSession() as session:
            async with session.get(
                f"{self.base_url}/models",
                headers=self.headers,
            ) as resp:
                return await resp.json()

    async def list_plugins(self) -> dict:
        """List installed plugins."""
        async with aiohttp.ClientSession() as session:
            async with session.get(
                f"{self.base_url}/plugins",
                headers=self.headers,
            ) as resp:
                return await resp.json()

    async def install_plugin(self, source: str, value: str) -> dict:
        """
        Install a plugin from file or URL.

        Args:
            source: "file" or "url"
            value: path to file or HTTPS URL
        """
        async with aiohttp.ClientSession() as session:
            async with session.post(
                f"{self.base_url}/plugins/install",
                json={"source": source, "value": value},
                headers=self.headers,
            ) as resp:
                return await resp.json()

    async def uninstall_plugin(self, plugin_id: str) -> dict:
        """Uninstall a plugin."""
        async with aiohttp.ClientSession() as session:
            async with session.delete(
                f"{self.base_url}/plugins/{plugin_id}",
                headers=self.headers,
            ) as resp:
                return await resp.json()

    async def plugin_action(self, plugin_id: str, action: str) -> dict:
        """
        Perform plugin action (enable, disable, restart).

        Args:
            plugin_id: Plugin identifier
            action: "enable", "disable", or "restart"
        """
        async with aiohttp.ClientSession() as session:
            async with session.post(
                f"{self.base_url}/plugins/{plugin_id}/action",
                json={"action": action},
                headers=self.headers,
            ) as resp:
                return await resp.json()

    async def get_task_history(self, limit: int = 100, status: str | None = None) -> dict:
        """Get task history."""
        params = {"limit": limit}
        if status:
            params["status"] = status

        async with aiohttp.ClientSession() as session:
            async with session.get(
                f"{self.base_url}/task-history",
                params=params,
                headers=self.headers,
            ) as resp:
                return await resp.json()

    async def listen_events(self):
        """Connect to WebSocket event stream and print events."""
        ws_url = "ws://127.0.0.1:8765/ws/events"
        print(f"[*] Connecting to {ws_url}")

        async with websockets.connect(ws_url) as ws:
            print("[+] Connected! Listening for events...")
            try:
                async for message in ws:
                    event = json.loads(message)
                    print(f"[EVENT] {event.get('type')}: {event.get('data')}")
            except asyncio.CancelledError:
                print("\n[-] Disconnected")


async def main():
    """Demo: connect to API and run sample commands."""
    client = RoaminControlClient()

    try:
        # Get status
        print("[*] Fetching status...")
        status = await client.get_status()
        print(f"[+] Status: {json.dumps(status, indent=2)}")

        # List models
        print("\n[*] Fetching models...")
        models = await client.list_models()
        print(f"[+] Models: {json.dumps(models, indent=2)}")

        # List plugins
        print("\n[*] Fetching plugins...")
        plugins = await client.list_plugins()
        print(f"[+] Plugins: {json.dumps(plugins, indent=2)}")

        # Get task history
        print("\n[*] Fetching task history...")
        tasks = await client.get_task_history(limit=10)
        print(f"[+] Tasks: {json.dumps(tasks, indent=2)}")

        # Listen for events (run for 5 seconds as demo)
        print("\n[*] Listening for events (5s timeout)...")
        try:
            await asyncio.wait_for(client.listen_events(), timeout=5)
        except asyncio.TimeoutError:
            print("[*] Event listener timeout (expected)")

    except ConnectionRefusedError:
        print("[!] Could not connect to Control API. Is it running?")
        print(f"    Expected: {client.base_url}")


if __name__ == "__main__":
    asyncio.run(main())
